### PR TITLE
CNF-11236: Apply extra manifests in the order specified by the `apply-wave` annotation

### PR DESCRIPTION
**Pull request by sudomakeinstall2**
Sort, group by and apply extra manifests with respect to `apply-wave`
annotation

Refactor sort, group by and load grouped manifests to use generics to
remove code duplications

Signed-off-by: Saeid Askari <saskari@redhat.com>

### CODE REVIEW

File: internal/extramanifest/extramanifest.go [hunk 1]
```diff
@@ -264,24 +265,30 @@ func (h *EMHandler) ExportExtraManifestToDir(ctx context.Context, extraManifestC
...
-	for i, manifest := range manifests {
-		gvk := manifest.GroupVersionKind()
-
-		fileName := fmt.Sprintf("%d_%s_%s_%s.yaml", i, gvk.Kind, manifest.GetName(), manifest.GetNamespace())
-		filePath := filepath.Join(toDir, ExtraManifestPath, fileName)
-		err = utils.MarshalToYamlFile(manifest, filePath)
-		if err != nil {
-			return fmt.Errorf("failed to marshal manifest %s to yaml: %w", manifest.GetName(), err)
+		for j, manifest := range manifestGroup {
+			fileName := fmt.Sprintf("%d_%s_%s.yaml", j+1, manifest.GetName(), manifest.GetNamespace())
```
Comment (Missxiaoguo) [inline internal/extramanifest/extramanifest.go:285]: gvk.Kind is removed from filename. Is it intentional?

File: internal/backuprestore/common.go [hunk 2]
```diff
@@ -399,11 +393,19 @@ func (h *BRHandler) ValidateOadpConfigmaps(ctx context.Context, content []lcav1a
...
-	restores, err := h.extractRestoreFromConfigmaps(ctx, configmaps)
+
+	for _, backup := range backups {
+		err := h.createBackupWithDryRun(ctx, backup)
+		if err != nil {
+			return err
+		}
+	}
+
+	restores, err := common.ExtractResourcesFromConfigmaps[*velerov1.Restore](ctx, configmaps, common.RestoreGvk)
```
Comment (Missxiaoguo) [inline internal/backuprestore/common.go:408]: Missing dryrun for restore CRs?

File: internal/common/manifest_helpers.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewBufferString(value), 4096)
+			for {
+				resource := unstructured.Unstructured{}
+				err := decoder.Decode(&resource)
+				if err != nil {
+					if errors.Is(err, io.EOF) {
+						// Reach the end of the data, exit the loop
+						break
+					}
+					return nil, fmt.Errorf("")
```
Comment (Missxiaoguo) [inline internal/common/manifest_helpers.go:81]: Should raise error with a message? i.e., failed to decode yaml

**Review by Missxiaoguo — COMMENTED** [review]
Thanks for the refactoring! Nice work!

File: internal/backuprestore/common.go (hunk 2 above)
Comment (sudomakeinstall2) [reply internal/backuprestore/common.go:398]: updated

File: internal/extramanifest/extramanifest.go (hunk 1 above)
Comment (Missxiaoguo) [reply internal/extramanifest/extramanifest.go:285]: Probably not a problem now as it's sorted file starting with number. It was added to ensure unique filename. But I still prefer to add it which helps to know the resource type by just looking at the name. The linux file length limit is long enough, so we don't need to worry about.

Comment (sudomakeinstall2) [reply internal/extramanifest/extramanifest.go]: updated

**Pull request merged**